### DISCUSSION TITLE
LP in matrix form

### DISCUSSION THREAD

**Discussion by riemanli** [optimatika/ojAlgo#495] [General, ojAlgo and related code]
I know ojAlgo has array/matrix data structure and it also supports LP. However, I can't find any example of using array/matrix in LP. Specifically, I am looking for an efficient way to solve the following problem setup.

```
minimize: c @ x
such that:
A_ub @ x <= b_ub
A_eq @ x == b_eq
lb <= x <= ub
```

Thanks.

Best,
Rieman

**Comment by apete** [comment]
The recommendation is to use `ExpressionsBasedModel` and don't worry about the underlying format.

If you absolutely want to use matrices to describe the the problem you can use `LinearSolver.newGeneralBuilder()` but then you loose the presolve functionality of `ExpressionsBasedModel`.

Also, the variable bounds specified using that builder are disregarded by the current solver implementation. That feature was added to support some ongoing work on a new linear solver. If you use `ExpressionsBasedModel` you don't need to worry about things like this.

**Reply by riemanli** [reply]
Anders,

Thanks for the prompt reply. If my understanding is correct, to use `ExpressionsBasedModel` we will need to use for-loops to set up variables and constraints. Please let me know if that's correct and if there is a better way to do it.

Best,
Rieman

**Reply by apete** [reply]
Yes, something like that. You cannot use the matrices directly. Refactor the code that creates the matrices to instead build an `ExpressionsBasedModel` instance.
